File: argon-plugin/src/main/scala/dev/argon/plugin/TubeLoader.scala
package dev.argon.plugin

import dev.argon.compiler.*
import dev.argon.options.OptionDecoder
import zio.*


trait TubeLoader[Ctx <: Context { type Env <: PluginEnv; type Error >: PluginError }] {
  type LibOptions[E >: PluginError]
  given libOptionDecoder[E >: PluginError]: OptionDecoder[E, LibOptions[E]]

  def load
  (context: Ctx)
  (tubeImporter: TubeImporter & HasContext[context.type])
  (libOptions: LibOptions[context.Error])
  : ZIO[context.Env & Scope, context.Error, ArTubeC & HasContext[context.type]]
}

File: plugin/api/java/src/main/java/dev/argon/plugin/api/TubeLoader.java
package dev.argon.plugin.api;

import dev.argon.plugin.api.options.OptionDecoder;
import dev.argon.plugin.api.tube.SerializedTube;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;

public interface TubeLoader<E extends Throwable, LibOptions> {
    @NotNull OptionDecoder<E, LibOptions> libOptionsDecoder();

    @NotNull SerializedTube<E> load(@NotNull TubeImporter<E> tubeImporter, @NotNull LibOptions libOptions) throws E, IOException, InterruptedException;
}
